slack_sdk/socket_mode/websocket_client/__init__.py

```diff
@@ -0,0 +1,192 @@
+        self.on_close_listeners = on_close_listeners or []
+
+    def connect(self) -> None:
+        def on_open(ws: WebSocketApp):
+            if self.logger.level <= logging.DEBUG:
+                self.logger.debug("on_open invoked")
+            for listener in self.on_open_listeners:
+                listener(ws)
+
+        def on_message(ws: WebSocketApp, message: str):
+            if self.logger.level <= logging.DEBUG:
+                self.logger.debug(f"on_message invoked: (message: {message})")
+            self.enqueue_web_socket_message(message)
+            for listener in self.on_message_listeners:
+                listener(ws, message)
+
+        def on_error(ws: WebSocketApp, error: Exception):
+            self.logger.error(
+                f"on_error invoked (error: {type(error).__name__}, message: {error})"
+            )
+            for listener in self.on_error_listeners:
+                listener(ws, error)
+
+        def on_close(ws: WebSocketApp):
+            if self.logger.level <= logging.DEBUG:
+                self.logger.debug("on_close invoked")
+            if self.auto_reconnect_enabled:
+                self.logger.info("Received CLOSE event. Going to reconnect...")
+                self.connect_to_new_endpoint()
+            for listener in self.on_close_listeners:
+                listener(ws)
+
+        old_session: Optional[
+            WebSocketApp
+        ] = None if self.current_session is None else self.current_session
+        self.current_session = websocket.WebSocketApp(
+            self.wss_uri,
+            on_open=on_open,
+            on_message=on_message,
+            on_error=on_error,
+            on_close=on_close,
+        )
+
+        def run_current_session():
+            self.current_session.run_forever(
+                ping_interval=self.ping_interval,
+                http_proxy_host=self.http_proxy_host,
+                http_proxy_port=self.http_proxy_port,
+                http_proxy_auth=self.http_proxy_auth,
+                proxy_type=self.proxy_type,
+            )
+
+        def monitor_current_session():
+            while True:
+                time.sleep(self.ping_interval)
+                try:
+                    if self.auto_reconnect_enabled and (
+                        self.current_session is None
+                        or self.current_session.sock is None
+                    ):
+                        self.logger.info(
+                            "The session seems to be already closed. Going to reconnect..."
+                        )
+                        self.connect_to_new_endpoint()
+                except Exception as e:
+                    self.logger.error(
+                        "Failed to check the current session or reconnect to the server "
+                        f"(error: {type(e).__name__}, message: {e})"
+                    )
+
+        self.current_app_executor.submit(run_current_session)
```

Comment:
not to block the current thread - WebSocketApp doesn't provide a way to establish WS connection without blocking the current thread.